flow/Histogram.h

```diff
@@ -35,64 +35,80 @@
 class Histogram;
 
 class HistogramRegistry {
-    public:
-        void registerHistogram(Histogram * h);
-        void unregisterHistogram(Histogram * h);
-        void logReport();
-    private:
-        std::map<std::string, Histogram*> histograms;
+public:
+	void registerHistogram(Histogram* h);
+	void unregisterHistogram(Histogram* h);
+	Histogram* lookupHistogram(std::string name);
+	void logReport();
+
+private:
+	std::map<std::string, Histogram*> histograms;
 };
 
-// TODO: This should be scoped properly for simulation (instead of just having all the "machines" share one histogram namespace)
-HistogramRegistry & GetHistogramRegistry();
+// TODO: This should be scoped properly for simulation (instead of just having all the "machines" share one histogram
+// namespace)
+HistogramRegistry& GetHistogramRegistry();
 
-class Histogram {
+class Histogram : public ReferenceCounted<Histogram> {
 public:
-    enum class Unit {
-        microseconds,
-        bytes
-    };
+	enum class Unit { microseconds, bytes };
 
-    Histogram(StringRef group, StringRef op, Unit unit) : group(group.toString()), op(op.toString()), unit(unit), registry(GetHistogramRegistry()) {
-        clear();
-        registry.registerHistogram(this);
-    }
+private:
+	Histogram(std::string group, std::string op, Unit unit)
+	  : group(group), op(op), unit(unit), registry(GetHistogramRegistry()) {
+		clear();
+		registry.registerHistogram(this);
+	}
 
-    ~Histogram() {
-        registry.unregisterHistogram(this);
-    }
+	static std::string generateName(std::string group, std::string op) { return group + ":" + op; }
 
-    inline void sample(uint32_t sample) {
+public:
+	~Histogram() { registry.unregisterHistogram(this); }
+
+	static Reference<Histogram> getHistogram(StringRef group, StringRef op, Unit unit) {
+		std::string group_str = group.toString();
+		std::string op_str = op.toString();
+		std::string name = generateName(group_str, op_str);
+		HistogramRegistry& registry = GetHistogramRegistry();
+		Histogram* h = registry.lookupHistogram(name);
+		if (!h) {
+			h = new Histogram(group_str, op_str, unit);
+		}
+		return Reference(h);
+	}
+
+	inline void sample(uint32_t sample) {
 #ifdef _WIN32
-        unsigned long index;
-        buckets[_BitScanReverse(&index, sample) ? index : 0]++;
+		unsigned long index;
+		buckets[_BitScanReverse(&index, sample) ? index : 0]++;
 #else
-        buckets[sample ? (31 - __builtin_clz(sample)) : 0]++;
+		buckets[sample ? (31 - __builtin_clz(sample)) : 0]++;
```

Comment:
fixed.  I left out the floor(), since careful readers will eventually find the part that compute the `LessThan#` for the histogram bucket attribute name.